### DISCUSSION TITLE
Exporting Zustand methods as a hook gives an error

### DISCUSSION THREAD

**Discussion by melihcoban** [pmndrs/zustand#1939] [Q&A] [answered]
Hi! First of all, I want to say that Zustand is a great library, I love it and want to thank everyone who is contributing to it. Today I came across https://tkdodo.eu/blog/working-with-zustand this blog post where it shows you can seperate methods in your store to an single object called "actions", and you can export it as a hook.

This is code in the blog: 

`const useBearStore = create((set) => ({
  bears: 0,
  fish: 0,
  // ⬇️ separate "namespace" for actions
  actions: {
    increasePopulation: (by) =>
      set((state) => ({ bears: state.bears + by })),
    eatFish: () => set((state) => ({ fish: state.fish - 1 })),
    removeAllBears: () => set({ bears: 0 }),
  },
}))`

`export const useBearActions = () =>
  useBearStore((state) => state.actions)`

This is my implementation:

```
const useTokenStore = create(
  persist<AuthState>(
    (set, get) => ({
      accessToken: undefined,
      refreshToken: undefined,

      actions: {
        setAccessToken: (accessToken: string) => {
          set({
            accessToken: { value: accessToken, expiry: new Date().getTime() + accessTokenExpiryTime, expired: false },
          });
        },
        setRefreshToken: (refreshToken: string) => {
          const cookieString = `"refresh_token"=${refreshToken}; max-age=${refreshTokenExpiryTime}; path=${refreshTokenPath}`;
          document.cookie = cookieString;
          set({
            refreshToken: { value: refreshToken, expiry: new Date().getTime() + refreshTokenExpiryTime, expired: false}
          });
        },
        clearTokens: () => {
          document.cookie = `refresh_token; Path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
          set({
            accessToken: undefined,
            accessTokenData: undefined,
            refreshToken: undefined
          });
        }
      }
    }),
    {
      name: 'tokens',
      storage: createJSONStorage(() => sessionStorage),
    }
  )
);
```

`export const useTokenActions = () => useTokenStore(state => state.actions);`

This is how I use it inside code:

```
function AuthenticationProvider({ children }: AuthenticationProviderProps) {

  const { setAccessToken, setRefreshToken, clearTokens } = useTokenActions();

  const authenticateUser = async (user: AuthenticationRequestDto): Promise<boolean> => {

    //// some business

    setAccessToken(response.data.accessToken);

    setRefreshToken(response.data.refreshToken);

    return true;
  };
```

When it comes to line `setAccessToken(response.data.accessToken);`, I get an error called "Uncaught (in promise) TypeError: setAccessToken is not a function". I could not figure out the issue, maybe the post is written for an old version of Zustand or maybe I am missing something. Thanks for the help!

**Comment by dbritto-dev** [comment, marked as answer]
@melihcoban you can't persist functions so you need to `partialize` your store to only persist `accessToken` and `refreshToken`

```ts
{
  name: 'tokens',
  partialize: (state) => ({ accessToken: state.accessToken, refreshToken: state.refreshToken }),
  storage: createJSONStorage(() => sessionStorage),
}
```
Keep me in the loop

**Reply by melihcoban** [reply]
@dbritto-dev  Thanks! Partializing the properties fixed the issue. However, I checked another store I am using and I am unintentionally persisting functions there without this problem. Maybe it happens here because the functions are in an action object. Another difference is that it uses local storage, but I don't think it makes much difference. I have another question.

**Reply by dbritto-dev** [reply]
@melihcoban sure, ping me whenever you want. BTW, try not to persist functions

**Reply by melihcoban** [reply]
@dbritto-dev I will be careful from now on, thanks for the heads up. I have an function for a state in a store where I need to check if the state is undefined or not.

```
getAccessToken: () => {
          const accessTokenState = get().accessToken;
          
          if (!accessTokenState) {
            return undefined;
          }

          const now = new Date();

          if (now.getTime() > accessTokenState.expiry) {
            set((state : AuthState) => ({accessToken: { value: accessTokenState.value, expiry: accessTokenState.expiry, expired: true }}));
            return get().accessToken;
          }

          return get().accessToken;
```

Is the code above is correct implementation, or is Zustand does not immediately updates the state like React? 

```
          if (now.getTime() > accessTokenState.expiry) {
            accessTokenState.expired = true;
            set((state : AuthState) => ({accessToken: { value: accessTokenState.value, expiry: accessTokenState.expiry, expired: true }}));
            return accessTokenState;
          }

          return accessTokenState;
```

Should I use the block-scoped variable for returning and set the state for future use? Another thing is that I know that access token is not undefined because I checked it before, therefore I should be able to use the spread operator to update one of its properties. 

```
set((state : AuthState) => ({ accessToken: { ...state.accessToken, expired: true }}));
```

However, because default type for accessToken is Token | undefined, I am getting an error. I can use the non-null assertion to make it work, but I am wondering if there is another solution. Thanks for the help! If it's wrong to continue here, I may open another discussion as these questions are completely unrelated to main one.

**Reply by dbritto-dev** [reply]
@melihcoban yeah, `getState` is not reactive  so returning the accessToken works
